arbor/morph/morphology.cpp

```diff
@@ -0,0 +1,208 @@
+#include <cmath>
+#include <iostream>
+#include <unordered_map>
+#include <utility>
+
+#include <arbor/math.hpp>
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/morphology.hpp>
+#include <arbor/morph/sample_tree.hpp>
+#include <arbor/morph/primitives.hpp>
+
+#include "algorithms.hpp"
+#include "util/span.hpp"
+#include "util/strprintf.hpp"
+
+namespace arb {
+
+namespace impl{
+
+std::vector<mbranch> branches_from_parent_index(const std::vector<size_t>& parents, const std::vector<point_prop>& props, bool spherical_root) {
+    using util::make_span;
+
+    const char* errstr_single_sample_root =
+        "A morphology with only one sample must have a spherical root";
+    const char* errstr_incomplete_cable =
+        "A branch must contain at least two samples";
+
+    auto npos = mbranch::npos;
+
+    if (!parents.size()) return {};
```

Comment:
Here and below: for std library containers, we can just use `empty()` to test for emptiness.

Reply:
Fixed for this line. The others `.size()` calls are on `sample_tree` and `mbranch` types, that don't implement `empty()`.

Reply:
Added an `empty` to `sample_tree`, and used.